Megaparsec/MonadParsec.lean

```diff
@@ -1,272 +1,290 @@
+    let el := Option.map ErrorItem.label (NEList.nonEmptyString l)
+    let f x s' hs :=
+      match el with
+      | .none => cok.2 x s' $ refreshLastHint hs .none
+      | .some _ => cok.2 x s' hs
+    let g x s' hs :=
+      eok.2 x s' $ refreshLastHint hs el
+    let ge err := eerr.2 $
+      match err with
+      | ParseError.trivial pos us _ => .trivial pos us (Option.option [] (fun x => [x]) el)
+      | _ => err
+    p xi s (cok.1, f) cerr (eok.1, g) (eerr.1, ge)
+
+  attempt p := fun xi s cok _ eok eerr =>
+    let forceBacktrack e _ := eerr.2 e s
+    p xi s cok (Consumed.mk, forceBacktrack) eok (Empty.mk, forceBacktrack)
+
+  lookAhead p := fun xi s _ cerr eok eerr =>
+    let forceBacktrack x _ _ := eok.2 x s [] -- TODO: why not forward hints?
+    p xi s (Consumed.mk, forceBacktrack) cerr (Empty.mk, forceBacktrack) eerr
+
+  notFollowedBy p := fun xi s _ _ eok eerr => do
     let o := s.offset
-    let pst := s.posState
-    let de := s.parseErrors
-    match stream.take1 input with
-      | .none => eerr (.trivial o (.some .eof) ps) s
-      | .some (c, cs) => match matcher c with
-        | .none =>
-            let us := (.some ∘ .tokens ∘ .uno) c
-            eerr (.trivial o us ps) s
-        | .some x => cok x (ParserState.State.mk cs (o + 1) pst de) []
-  tokens A matcher chunk := Parsec.ParsecT.mk $ fun B s₀ cok _ eok eerr =>
+    let y : (Chunk β × σ) ← Straume.take1 α s.input
+    let c2e := ErrorItem.tokens ∘ NEList.uno
+    let subject : ErrorItem β := match y.1 with
+    -- TODO: Here and in many other places, we have two branches that are the same because Parsec doesn't care about .fin vs .cont
+    -- TODO: Perhaps, we should use Terminable and extract values
+    | .nil => ErrorItem.eof -- If by the time we call notFollowedBy the stream is empty, treat it as eof.
+    | .cont c => c2e c
+    | .fin (c, _reason) => c2e c -- It's ok to work with .fin, because we never consume.
+    let ok _ _ _ := eerr.2 (.trivial o subject []) s
+    let err _ _ := eok.2 PUnit.unit s []
+    p xi s (Consumed.mk, ok) (Consumed.mk, err) (Empty.mk, ok) (Empty.mk, err)
+
+  withRecovery φ p := fun xi s cok cerr eok _ =>
+    let err (fHs := (hs₀ β σ E)) e sFail :=
+      let ok ψ := fun x s' _hs => ψ x s' (fHs s' e)
+      let err _ _ := cerr.2 e sFail
+      (φ e) xi sFail (cok.1, ok cok.2) (Consumed.mk, err) (eok.1, ok eok.2) (Empty.mk, err)
+    p xi s cok (Consumed.mk, err) eok (Empty.mk, err $ hs' β σ E)
+
+  observing p := fun xi s cok _ eok _ =>
+    let err (fHs := (hs₀ β σ E)) e s' := cok.2 (.left e) s' (fHs s' e)
+    p xi s (cok.1, cok.2 ∘ .right) (Consumed.mk, err) (eok.1, eok.2 ∘ .right) (Empty.mk, err (hs' β σ E))
+
+  eof := fun _ s _ _ eok eerr => do
+      let y : (Chunk β × σ) ← Straume.take1 α s.input
+      let err c := eerr.2 (.trivial s.offset (.some $ ErrorItem.tokens $ NEList.uno c) ([.eof])) s
+      match y.1 with
+      | .nil => eok.2 PUnit.unit s []
+      | .cont c => err c
+      | .fin (c, _) => err c
```

Comment:
Here's the problem: semantics of MonadParsec is "terminate when the stream is empty, no matter why", whereas semantics of Straume is "terminate when the stream is interrupted for some reason, remembering the reason".

Since MonadParsec is termination-reason-agnostic, `.nil` is the proper marker of the end of a parse.

Reply:
I will write a test for this. I like the idea of checking against files, because this way we can see that rewinds happen properly with old inputs. But I think there are no bugs and takeP 4 will give you "abcd", consuming `.fin 'd'`, and then in `eof`, `take1` will get you a `nil` and thus `eof` will succeed.

I'll check it in `main` though.